core/server/api_container/server/service_network/default_service_network.go

```diff
@@ -1791,3 +1878,102 @@ func renderTemplateToFile(templateAsAString string, templateData interface{}, de
 	}
 	return nil
 }
+
+func waitUntilAllTCPAndUDPPortsAreOpen(
+	ipAddr net.IP,
+	ports map[string]*port_spec.PortSpec,
+) error {
+
+	var portCheckErrorGroup errgroup.Group
+
+	for _, portSpec := range ports {
+		wrappedWaitFunc := func() error {
+			return waitUntilPortIsOpenWithTimeout(ipAddr, portSpec)
+		}
+		portCheckErrorGroup.Go(wrappedWaitFunc)
+	}
+	//This error group pattern allow us to reject early if at least one of the ports check fails
+	//in this opportunity we want to reject early because the main user pain that we want to handle
+	//is a wrong service configuration
+	if err := portCheckErrorGroup.Wait(); err != nil {
+		return stacktrace.Propagate(err, "An error occurred while waiting for all TCP and UDP ports to be open")
+	}
+
+	return nil
+}
+
+func waitUntilPortIsOpenWithTimeout(
+	ipAddr net.IP,
+	portSpec *port_spec.PortSpec,
+) error {
+	// reject early if it's disable
+	if !portSpec.GetWait().GetEnable() {
+		return nil
+	}
+	timeout := portSpec.GetWait().GetTimeout()
+
+	var (
+		ticker                  = time.NewTicker(timeout)
+		startTime               = time.Now()
+		finishTime              = startTime.Add(timeout)
+		shouldContinueInTheLoop = true
+		retries                 = 0
+		err                     error
+	)
+
+	time.Sleep(portSpec.GetWait().GetInitialDelay())
+
+	for shouldContinueInTheLoop {
+		select {
+		case <-ticker.C:
```

Comment:
Oh wow, that's an interesting way to use a Ticker here! What about:
```
ticker = time.NewTicker(waitForPortsOpenRetriesDelayMilliseconds * time.Millisecond)
for {
    if time.Now().After(finishTime) {
        // throw a TIMEOUT error
    }
    //run your check and break the loop if it's successful, or if a hard failure is hit
    <-ticker.C // block until the next tick
}
```
I personally find it more in sync with the idea of a Ticker (which ticks at a fixed rate). Here you're using a ticker only to grab the first tick, it's a bit sad.
And, I believe you have one less indentation level here, since you don't have the `select`

Reply:
good suggestion! I'll do the change